### PR TITLE
Coalesce requests to KV

### PR DESCRIPTION
**Pull request by mendess**
This PR is best viewed commit by commit.

- Add get or insert method to kv (this commit is setup, this method will later being coalesced)
- Abstract away the internals of kv connection
- rename get_impl to get_internal
- Add request coalescer to kv (this is the important commit to review)

### CODE REVIEW

File: crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+use std::{
+    any::Any,
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    marker::PhantomData,
+    sync::Mutex,
+};
+use tokio::sync::watch::{self, Receiver, Sender};
+
+/// For coalescing to work correctly the key the key we use for lookup has to be simultaneously:
```
Comment (armfazh) [inline crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs:27]:
```suggestion
/// For coalescing to work correctly the key we use for lookup has to be simultaneously:
```

File: crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+    #[cfg(feature = "test-utils")]
+    shutting_down: std::sync::atomic::AtomicBool,
+}
+
+impl RequestCoalescer {
+    #[cfg(feature = "test-utils")]
+    pub async fn reset(&self) {
+        self.shutting_down
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+        let recievers = self
```
Comment (armfazh) [inline crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs:60]:
```suggestion
        let receivers = self
```

File: crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+            #[cfg(feature = "test-utils")]
+            assert!(
+                !self.shutting_down.load(std::sync::atomic::Ordering::SeqCst),
+                "RequestCoalescer is shutting down"
+            );
+
+            let op_key = OperationKey {
+                op_id: operation_id::of::<F, Fut, R, E>(),
+                key: key.clone(),
+            };
```
Comment (armfazh) [inline crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs:98]: can the `op_key` variable be defined before the loop?

File: crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+            Self::Canceled => Self::Canceled,
+            Self::Error(e) => Self::Error(e.clone()),
+        }
+    }
+}
+
+type CoalescedValue =
+    Result<Option<Marc<dyn Any + Send + Sync>>, CoalesceError<dyn Any + Send + Sync>>;
+
+/// The send half to be used by the `executor` to send the result of it's computation to the
```
Comment (armfazh) [inline crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs:151]:
```suggestion
/// The send have to be used by the `executor` to send the result of it's computation to the
```

**Review by armfazh — APPROVED** [review]
Nice functionality, changes are ok.

File: crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs (hunk 4 above)
Comment (mendess) [reply crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs:151]: This is actually correct. A channel has two parts, the sending part and the receiving part, commonly referred to as the sending and receiving halves

File: crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs (hunk 3 above)
Comment (mendess) [reply crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs:98]: That results in an additional clone when passing it to the ResponseSender

**Pull request merged**